docs/low-code-spark/extensibility/gem-builder.md

```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
+        )
+        .addColumn(
+          StackLayout(height = Some("100%"))
+            .addElement(TitleElement("Filter Condition"))
+            .addElement(
+              Editor(height = Some("100%"))
+                .withSchemaSuggestions()
+                .bindProperty("condition.expression")
+            ),
+          "5fr"
+        )
+    )
+
+  def validate(component: Component)(implicit context: WorkflowContext): List[Diagnostic] = {
+    val diagnostics =
+      validateSColumn(component.properties.condition, "condition", component)
+    diagnostics.toList
+  }
+
+  def onChange(oldState: Component, newState: Component)(implicit context: WorkflowContext): Component = {
+    val newProps = newState.properties
+    val portId = newState.ports.inputs.head.id
+
+    val expressions = getColumnsToHighlight(List(newProps.condition), newState)
+
+    newState.copy(properties = newProps.copy(columnsSelector = expressions))
+  }
+
+  class FilterCode(props: PropertiesType)(implicit context: WorkflowContext) extends ComponentCode {
+
+    def apply(spark: SparkSession, in: DataFrame): DataFrame = {
+      val out = in.filter(props.condition.column)
+      out
+    }
+
+  }
+
+}
+
+```
+
+</TabItem>
+</Tabs>
+````
+
+### Parent Class
+
+Your Gem class needs to extend a parent class from which it inherits the representation of the overall Gem. This includes the UI and the logic.
+For transform Gems, you need to extend `ComponentSpec` (like in the example above), and for Source/Target Gems you need to extend `DatasetSpec`. We will see the difference between the two at the end.
+
+First thing you give after this is the name and category of your Gem, `"Filter"` and `"Transform"` in this example.
+
+Another thing to note here is `optimizeCode`. This flag can be set to `True` or `False` value depending on whether we want the Prophecy Optimizer to run on this code to simplify it.
+In most cases, it's best to leave this value as `True`.
+
+````mdx-code-block
+<Tabs>
+
+<TabItem value="py" label="Python">
+
+```py
+class Filter(ComponentSpec):
+name: str = "Filter"
+    category: str = "Transform"
+    def optimizeCode(self) -> bool:
+        return True
+```
+</TabItem>
+
+<TabItem value="scala" label="Scala">
+
+```scala
+object Filter extends ComponentSpec {
+val name: String = "Filter"
+val category: String = "Transform"
+override def optimizeCode: Boolean = true
+```
+</TabItem>
+</Tabs>
+
+````
+
+### Properties Classes
+
+There is one class (seen here as `FilterProperties`) that contains a list of the properties to be made available to the user for this particular Gem. Think of these as all the values a user fills out within the template of this Gem, or any other UI state that you need to maintain (seen here as `columnsSelector` and `condition`).
+
+:::caution
+
+The content of these `Properties` classes is persisted in JSON and stored in Git, so be careful not to keep any information in them that could be a potential security risk.
```

Comment:
The content of these `Properties` classes is persisted in JSON and stored in Git. 

(omit the remainder of the sentence)